### PR TITLE
 Introduce "collection" projects for better usage of hierarchical view #2041 

### LINKED ISSUE #2041: Introduce "collection" projects for better usage of hierarchical view
### Current Behavior:
After implementation of #84, grouping different versions or environments into one project is a great feature and improves the overview in the project list. But it has one drawback: You now need to expand a project to see the current issue count, if the project is just a "meta"-level to group versions or sub-projects.

### Proposed Behavior:
When selecting a project Classifier, there could be an option like "Project Collection". This would be applied to intermediate versions like "Project X - Android App" (containing multiple versions as children) or "Project X - Server" (containing multiple microservices below). 

If "Project collection" is selected, additional configuration shall be shown:
A dropdown to select the "Metrics type", which can be one of:
- Latest uploaded BOM
- Sum of direct children metrics
- Specific child

The UI and API should then prevent upload of BOM and managing components manually for such collection project, and instead show the numbers of its children, according the configuratio.  This would give better overview in the UI and allow additional use cases like https://github.com/DependencyTrack/dependency-track/issues/657, e.g. by later replacing the component view for collection projects with a metrics view.

The options logic should be:

**Latest uploaded BOM**
The numbers of the (direct) child for which the BOM was uploaded last shall be shown on the parent. This would be helpful for versioned software, so you can see always the latest versions data on the parent.

**Sum of direct children metrics**
The numbers of all (direct) children shall be summed up and taken for the parent. This could be interesting, e.g. if you have a meta level "Cloud Project X - Productive environment", with microservices as children, so you see the summary of the complete production environment

**Specific child**
If selected, offers another input field to select a specific children somehow*. This can be useful in cases where e.g. "Cloud Project X" has direct children for "develop", "qa", "productive" (see above example), so you can always show the "productive" childs numbers on the parent, to focus on them in the views. PROBLEM: If you want to select a child, you cannot do this while creation, and also it would introduce a circular reference. Possible alternative: Define a tag name instead in a textbox input, which is then used to label the child. The first child found with that tag would then be taken.

When calculating numbers with a child, which also is a collection, the numbers calculated for this collection should be taken, not all of its children, like this you can configure yourself a nice hierarchy. 

**Performance**
Live calculation would be a problem for a big project structure. Therefore, the existing vulnerability analysis task should as last step get another step to re-calculate collection metrics. If we have some kind of information about which projects changed during the analysis, we could possibly limit the re-calculation to only trees in which numbers changed.
Additionally, when adding/deleting new sub-projects, changing a sub-projects parent, changing collection configuration or uploading a BOM, (changing a tag if it matches the parents tag configuration if this approach is used), the metrics for the hierarchy above the changed project would have to be analyzed instantly. The results need to be cached.

### Considerations to make:
- Need to ensure when calculating metrics for the portfolio, the collection projects are not count, otherwise we would count duplicates
- What if someone doesn't have access to all children projects, but only to a subset? I think if this is the case, he can see the full metrics, even if he only sees a subset. I can think of some controlling having access to top level metrics while not seeing into detailed projects. Also other solutions would not work with cached data.

### LINKED ISSUE #2410: Parent shows not the sum of the childs
### Current Behavior

Since last version it is possible to have an parent project with sub projects.
I should expect that the parent will sum the results of the childs.

It can also be possible that it was already know on the backlog but not yet finished,

### Steps to Reproduce

1. Have a parent project
2. Have 2 child projects
project_1 has 1 medium finding
project_2 has 1 low finding

### Expected Behavior

Parent shows everything empty and zero, but should expect parent project has 1 medium and 1 low.

### Dependency-Track Version

4.6.x

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Database Server Version

_No response_

### Browser

Google Chrome

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

### PR DESCRIPTION
**Pull request by rkg-mm**
### Description
This change introduces logic for "collection projects". Those are basically projects used as parent for other projects that shall not hold any own component or vulnerability data, but instead get calculated from child projects using different configurable aggregation logics.

Corresponding Frontend PR with Screenshots: https://github.com/DependencyTrack/frontend/pull/658

### Addressed Issue
resolves #2041 
resolves #2410

### Checklist
- [x] I have read and understand the [contributing guidelines](../CONTRIBUTING.md#pull-requests)
- [ ] ~This PR fixes a defect, and I have provided tests to verify that the fix is effective~
- [x] This PR implements an enhancement, and I have provided tests to verify that it works as intended
- [x] This PR introduces changes to the database model, and I have added corresponding [update logic](https://github.com/DependencyTrack/dependency-track/tree/master/src/main/java/org/dependencytrack/upgrade)
- [x] This PR introduces new or alters existing behavior, and I have updated the [documentation](https://github.com/DependencyTrack/dependency-track/tree/master/docs/_docs) accordingly

### CODE REVIEW

**Comment by cheapshot2000** [comment]
Has there been any updates/progress on the feature in completion of the 4.11 release?

**Comment by TWpgo** [comment]
Looking forward for this feature 👍

**Comment by rkg-mm** [comment]
I messed up the branch trying to rebase onto master to resolve 1 conflicting file... ending up with >1300 changed files now that already are changed in master. no idea what git is doing... will try to fix it but I'm struggling a bit how to get back to a clean state

**Comment by rkg-mm** [comment]
Warning: I fixed a lot of conflicts and did not have time to test it. Will try to find time tomorrow to test it.

**Comment by rkg-mm** [comment]
@nscuro I wonder why the unit test still fails. I don't see a change that could influence the parent of a project not being serialized?

**Comment by nscuro** [comment]
@rkg-mm FWIW there can be subtle changes in how serialization libraries (i.e. Jackson) handle certain fields, when those libraries are updated. Or how the persistence framework fetches fields implicitly.

**Comment by rkg-mm** [comment]
Yea so, the PRs backend & frontend seem to be working well. But I have no clue about this unit test and will need some help here. The test explicitly tests the now not working condition, so I guess its important that it works, but I don't see a change that causes the failure.

```
 Error:  Tests run: 40, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 74.367 s <<< FAILURE! - in org.dependencytrack.resources.v1.ProjectResourceTest
Error:  patchProjectParentTest(org.dependencytrack.resources.v1.ProjectResourceTest)  Time elapsed: 1.483 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: 
JSON documents are different:
Different keys found in node "", missing: "parent", expected: 
<{"active":true,"collectionLogic":"NONE","name":"DEF","parent":{"name":"GHI","uuid":"${json-unit.matches:parentProjectUuid}","version":"3.0"},"properties":[],"tags":[],"uuid":"${json-unit.matches:projectUuid}","version":"2.0"}> 
but was: 
<{"active":true,"collectionLogic":"NONE","name":"DEF","properties":[],"tags":[],"uuid":"ba033fa5-aa20-44ae-85de-6a33e5cf36a3","version":"2.0"}>

	at org.dependencytrack.resources.v1.ProjectResourceTest.patchProjectParentTest(ProjectResourceTest.java:733)
```

**Comment by nscuro** [comment]
> @rkg-mm But I have no clue about this unit test and will need some help here. The test explicitly tests the now not working condition, so I guess its important that it works, but I don't see a change that causes the failure.

I debugged the area in question multiple times now and still don't quite understand how this is happening. The parent is part of the response when I comment out these lines: https://github.com/DependencyTrack/dependency-track/pull/3258/files#diff-75cd29a0a085d84dff3cb4794242e0d1b400385f1acfae172346c3e6780629b0R582-R590

Something about those lines causes `parent` to be unloaded. But they're literally just calling a setter, so I don't quite get why. The test succeeds when adding `project.getParent();` after the lines mentioned above. This kind of "force loading" is used in other areas of the code base. I don't like it, but in this case I think it's OK to do it.

File: src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java:48]: I think its pretty important for versioned software.
Only alternative I could think of:
One needs logic in CI scripts to identify the latest SemVer version in Dtrack after creating a version and must auto-manage tags to always append a "highest-version" tag or similar to the latest version on its own and auto-remove it for all other versions. Then you can use the "aggregate by tag" logic to achieve the same. But this requires quite some logic in CI builds.
Or could we at least offer an API to identify the latest SemVer version of a project? That would not be called as often as Metrics update, hence doing it outside SQL would be ok. Then it would be a step easier to implement above logic. 

Or even better: one could enable an optional feature that lets the project management APIs auto-tag the latest SemVer version automatically whenever projects are added/removed/reversioned. 
Would this be an option? Then we don't lose any comfort AND the info could even be reused for other potential features, while not slowing down or making more complicated the metrics update.

File: src/main/java/org/dependencytrack/tasks/metrics/ProjectMetricsUpdateTask.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/tasks/metrics/ProjectMetricsUpdateTask.java:122]: @nscuro  
I would assume the latest update would then have the latest data, even if running more or less in parallel. If we really would face such issue in rare cases, it would probably fix itself with the next portfolio metric update. or?

Regarding Performance:
What we could do as quick improvement is adding a condition in line 119 to only trigger the parent update when the counters changed. Would need to overthink again if there was a reason why I didn't do this or if I just forgot this.
Usually, the counters should not change too often I would assume, or?

File: src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java:550]: That's correct, but also it would be an intended change. Idk what could be done about this except add such change to the logfile, but that would only be helpful for deep analysis of such issue.

Comment (rkg-mm) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java:550]: Agreed, I should add this.

**Comment by rkg-mm** [comment]
oh come on.. Idk what IntelliJ is doing but whenever I rebase this branch on your master I end up having all changes from master in my PR again. What am I doing wrong here? I don't get it...

**Comment by nscuro** [comment]
@rkg-mm How do you do the rebase? I usually run `git pull --rebase upstream master` in the command line and then use IntelliJ to resolve any conflicts in case there are any.

I never had this particular problem happen, but can imagine IntelliJ might do some unexpected things when merging/rebasing via UI.

**Comment by rkg-mm** [comment]
In IntelliJ UI I select the upstream master and select "Rebase (my branch) onto master" then push the changes to my remote. The strange thing happens here already, as suddenly my own remote has changes not recognized by my local branch and needs to either be merged (which I tried before and ended up with double changes on same files) or I select "rebase anyway" which avoids double merging. But in both cases I still end up with all those unwanted changes again. Maybe I should try the console attempt next time...

File: src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java]: Done

File: src/main/java/org/dependencytrack/tasks/metrics/ProjectMetricsUpdateTask.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/tasks/metrics/ProjectMetricsUpdateTask.java]: I added a check if metrics of current project changed and only trigger an update of parent in this case. At least reduces the calculation effort.

**Comment by nscuro** [comment]
@rkg-mm We have to schedule this for 4.12 to allow us for sufficient time to review and test.

**Comment by rkg-mm** [comment]
> @rkg-mm We have to schedule this for 4.12 to allow us for sufficient time to review and test.

no problem, I understand. Lets just hope the release cycle is a bit shorter this time :D

**Comment by cheapshot2000** [comment]
> > @rkg-mm We have to schedule this for 4.12 to allow us for sufficient time to review and test.
> 
> no problem, I understand. Lets just hope the release cycle is a bit shorter this time :D

Is there a set release cadence?  or other logic for determining when releases are completed?

**Comment by nscuro** [comment]
> Is there a set release cadence? or other logic for determining when releases are completed?

There is not at the moment. We originally planned to do monthly releases, but ended up stretching that (by far...) yet again. A problem is that the core team has certain items they want (and have to work on) in a given release, while said core team also needs to review and test community contributions. It's a luxury problem, but it means that more time is needed to meet both our own, and the community's expectations.

This particular PR was planned for inclusion in 4.11 for a long time, but it also touches on the very core of how DT behaves, in turn requiring more review and testing than other, smaller PRs. And of course contributors also don't always have time to answer our annoying requests. :)

We're still in the process of figuring out how to best plan and size releases. Going forward, we're labelling issues with T-shirt sizes, according to their effort. Aim being to limit the number of high-effort items per release, enabling faster cycles.

File: src/main/java/org/dependencytrack/tasks/metrics/ProjectMetricsUpdateTask.java
Comment (nscuro) [reply src/main/java/org/dependencytrack/tasks/metrics/ProjectMetricsUpdateTask.java]: The check you added seems to work well in my initial testing. 👍 

However I think a test with a large portfolio would be justified to ensure the impact is indeed limited.

File: src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java
Comment (nscuro) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java:48]: > Or could we at least offer an API to identify the latest SemVer version of a project?

When dealing with versioned software like this, wouldn't it be a fair assumption that BOMs of older versions would have been uploaded earlier than those of newer versions?

At the very least, the latest version should be somewhere among the most recently uploaded BOMs, no? So you could use this endpoint in CI to pull the last, say, `5` projects with name `foo` and then do the version comparison as needed:

```shell
curl -H 'X-Api-Key: <API_KEY>' \
    'http://localhost:8080/api/v1/project?name=foo&sortName=lastBomImport&sortOrder=desc&offset=0&limit=5'
```

To be clear, I'm not strictly against supporting SemVer in DT, I'm just trying to avoid us steering into a maintenance mess once people start asking for more versioning schemes.

File: src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
Comment (nscuro) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java]: Fair.

Naive idea: We could add a *Collection logic change* flag to `ProjectMetrics`. We could use that in the metrics graph, in some similar way to what you can do in Grafana with annotations:

image

File: src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java:48]: > When dealing with versioned software like this, wouldn't it be a fair assumption that BOMs of older versions would have been uploaded earlier than those of newer versions?

Not necessarily.
Lets say you have 4 major versions
1.x.x, 2.x.x, 3.x.x, 4.x.x
and you support e.g. the last 3 of those due to business contracts (>=2.x.x) additionally you have active development on the latest version (4.x.x). In this case it can easily happen your upload order is for a while (first to latest):
4.1.0
4.1.1 (bugfix)
3.4.1 (backport of bugfix)
2.3.1 (backport of bugfix)
2.3.2 (additional fix in older version that doesn't apply to newer version)
...

> At the very least, the latest version should be somewhere among the most recently uploaded BOMs, no? So you could use this endpoint in CI to pull the last, say, 5 projects with name foo and then do the version comparison as needed:
Likely, yes. I would also assume for our current use cases 5 would be enough. However, idk if that would be the case for everyone. Special cases (e.g. newly using Dtrack and importing existing SBOMs of a lot of SW versions) also not considered.

Not sure whats the best strategy here...

File: src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java]: I like that idea, but at least on a first look the frontend chart lib doesn't support this. not sure how difficult this would be to add on top of it. Any experience with that? I never used that component.
From data perspective in backend it should not be too complicated.

Comment (nscuro) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java]: I don't have *too* strong opinions on *how* we highlight it, but I do think that highlighting it one way or the other would be good.

Tooltip on it's own might be too hidden, but styling the datapoints sounds good.

File: src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java
Comment (nscuro) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java:48]: > I would also assume for our current use cases 5 would be enough. However, idk if that would be the case for everyone.

Nothing would prevent users to traverse the project list to the very end in order to determine the "latest" version.

> Not sure whats the best strategy here...

My suggestion would be to investigate **if** and **how** the use case can be addressed with tools we already have available, including pulling all project versions and doing the "latest" determination on the client-side (i.e. CI scripts).

Based on that, figure out what the pains / limitations are with that approach, and how a server-side implementation could help resolve that. Are the identified limitations show-stoppers or merely annoyances?

If we are **really** sure we need a server-side implementation, we will need something that covers more than vanilla SemVer: Some organizations will version their software according to Maven, others may use .NET semantics, and so on. We also need to consider that organizations may track software they did not build (i.e. container images from vendors), and thus don't have control over versioning. The `vers` specification lists a few versioning schemes here: https://github.com/package-url/purl-spec/blob/version-range-spec/VERSION-RANGE-SPEC.rst#some-of-the-known-versioning-schemes

We don't have to cover all cases from the very beginning, but the solutions should at least be extensible such that more can be added in the future.

---

More generally, my recommendation would be to decouple the version-awareness from this PR, as to not block this from getting merged.

**Comment by msymons** [comment]
Re-assigning to 4.13 milestone in order to reduce the pressure on contributors to get this finished whilst allowing v4.12.0 to be released quicker.

File: src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java]: @nscuro would you be fine with creating a new issue to add this visibility later as further enhancement, and for now rely on logile for visibility in case someone wonders? we could also add a hint in the project edit UI when someone changes the logic, that informs him of this for now?

File: src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java:48]: Ok since I somehow want to get this feature out and have limited time, lets find a way for now.
In my opinion being able to auto-point to the latest version would be very important. for me this would be like half the projects that should rely on this.

However, I see your point that an overall and more flexible versioning solution makes sense. But, I currently can't take care of another big topic due to time constraints.
So either we take the solution suggested now as a temporary solution and state in the documentation that there are plans to add more flexibility later, allowing us to replace it in a future version, or we remove this. Possibly could replace it for now with a "LATEST_CHILD_BOM_UPLOADED", which would purely rely on the timing of the uploaded BOM (however this could produce flaky/unstable results if multiple versions are developed in parallel like a new major and a patch minor version). 
Using my current solution as temporary solution shouldn't block further replacement in my opinion, and would possibly allow us to receive feedback about requirements of others for this topic, especially if we link an improvement github Issue in the docs directly.

What do you think @nscuro? What do others think?

Comment (rkg-mm) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java:48]: Sounds like a good compromise to keep it simple. will you implement that in a separate PR first?

Comment (nscuro) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java:48]: Yeah this should be a separate change I think. In the meantime you can either omit the "highest version" logic in this PR, or implement it with dummy logic until the other PR lands.

File: src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
Comment (nscuro) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java]: Logging is probably fine for an MVP implementation, yes. But ultimately we will really need some kind of indicator in the UI.

File: src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java]: I have removed the highestsemverchild logic and replaced it with a dummy aggregating children having a future isLatest property set. I created https://github.com/DependencyTrack/dependency-track/issues/4148 to get this flag done. Once https://github.com/DependencyTrack/dependency-track/issues/4148 is implemented, the logic can be activated and only requires adding a test case, can be done as part of https://github.com/DependencyTrack/dependency-track/issues/4148, I added some notice in that issue that it shall be done when doing this.

File: src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java]: @nscuro  I found a way to show it in the metrics chart. Only issue is that the line gets invisible when hovered. No Idea why. But I think this is fine as it is, looks like its intended :D

Comment (rkg-mm) [reply src/main/java/org/dependencytrack/persistence/ProjectQueryManager.java]: Ok found the reason for the visibility issue and fixed it. Also extended it to the other charts now. Looks like this:

image

**Comment by rkg-mm** [comment]
@nscuro apart from the one proposal in the UI PR which I added some question for, this should be finished now. Can you review it again please?

**Comment by rkg-mm** [comment]
> Nice to see this PR moving forward. We run a previous version of the PR in our experimental setup. Will update with the latest version and provide feedback.
> 
> Some minor thing that we noticed and that would be great if that could be changed or supported as well: when modeling your products in a parent / child relationship like that:
> 
> image
> 
> You are forced to set a classifier for the parent project, though it is purely a project container and the classifier does not make any sense in this context. If the classifier could be made optional or better reflect the concept of a project container, it would be nice and avoid confusion when looking at the project hierarchy.

So basically you want to have an option which marks the parent as a collection project, hence removes the component uploads etc. but does not calculate any values based on children? what should be displayed inside the parent then? nothing?

**Comment by netomi** [comment]
so the behavior is just fine as it is right now. It feels just odd that when modeling your projects with dependency track you are forced to enter some data for a project whose only purpose is to act as a container for nested projects.

Right now you have to select at least a classifier afaict, so in our case (see the screenshot) we have selected Application, though it is not an Application, but the project that consists of various software products. When selecting that project, you should see all vulnerabilities of child projects as done right now.

This was discussed yesterday in the community call and its a really small thing but people asked me to raise it in this ticket as well.

File: src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java:30]: Docs added. However, since I'm stuck to Windows and there is no build script for the docs in windows, I could not really test if this integrates properly. Can you check this?

File: src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java
Comment (rkg-mm) [reply src/main/java/org/dependencytrack/model/ProjectCollectionLogic.java]: Enabled now the logic for AGGREGATE_LATEST_VERSION_CHILDREN and added test for it.

**Comment by rkg-mm** [comment]
@nscuro 
From my point of view this PR is good to go now, would be great if you could review this before I have a lot of conflicts to fix again ;-)
Are you fine with making the classifier optional (always or only in case of collections?) as requested by @netomi ? I missed the community meeting and the recording is not there yet, so I cannot check what was said there.

**Comment by rkg-mm** [comment]
> @nscuro From my point of view this PR is good to go now, would be great if you could review this before I have a lot of conflicts to fix again ;-) Are you fine with making the classifier optional (always or only in case of collections?) as requested by @netomi ? I missed the community meeting and the recording is not there yet, so I cannot check what was said there.

I added a NONE classifier now after watching the community meetings on youtube.
I furthermore noticed that Frontend does not support all classifiers supported by backend so far, so I also added support for the missing classifiers in Frontend.

**Comment by netomi** [comment]
fyi: a package of this PR is available at https://github.com/users/netomi/packages/container/dtrack-apiserver/289607235?tag=4.13.0-SNAPSHOT

**Comment by rkg-mm** [comment]
> * tags doesn't really need to be in direct children for me. We could have a project _images_ who contains all images used by different projects/applications and only create a _meta view_ of an project/application with all dependencies based on tag. To avoid duplicated data?
Looping through the complete child-subtrees (and on updates of a project through all parents to see if there is any metrics to update) would have a significant performance effect in large instances, which is why I limited possible rules to direct children only.

> * limit the propagation of metrics to one level is made on purpose?
Unless I made a mistake (but I think I tested this) the propagation should go to the direct parent (due to above reasons). But once the parents metrics change, this will trigger an update of the parent's parent, if this also is a calculated project, and so on, so you can build up whole calculated trees this way. You can see the logic in this screenshot https://raw.githubusercontent.com/DependencyTrack/dependency-track/f14fc9a005743f74b5a8ae5812792eac3b956cc5/docs/images/screenshots/collection-projects-structure.png

**Comment by nscuro** [comment]
@rkg-mm I raised https://github.com/rkg-mm/dependency-track/pull/1 to resolve the merge conflicts, please have a look.

**Comment by nscuro** [comment]
@rkg-mm Had a few more suggestions / comments and just went ahead and raised a PR for them instead of just commenting: https://github.com/rkg-mm/dependency-track/pull/2

Please have a look and let me know if the proposed changes make sense.

**Review by nscuro — APPROVED** [review]

**Pull request merged**
